File: scrapper/scrapeTender.js
import puppeteer from "puppeteer";

export async function scrapeTender(url) {
  const browser = await puppeteer.launch({
    headless: true,
    args: ["--no-sandbox", "--disable-setuid-sandbox"]
  });

  const page = await browser.newPage();
  await page.goto(url, { waitUntil: "networkidle2" });

  await page.waitForSelector(".doc-title");

  const data = await page.evaluate(() => {
    const clean = t => t?.replace(/\s+/g, " ").trim();

    const metaEls = document.querySelectorAll(".meta p");

    const sections = {};
    document.querySelectorAll("section").forEach(sec => {
      const title = clean(sec.querySelector("h2")?.innerText);
      const content = clean(sec.innerText.replace(title, ""));
      if (title) sections[title] = content;
    });

    return {
      project_name: clean(document.querySelector(".doc-title")?.innerText),
      issued_by: clean(metaEls[0]?.innerText.replace("Issued By:", "")),
      rfp_reference: clean(metaEls[1]?.innerText.replace("RFP Reference No.:", "")),
      date_issued: clean(metaEls[2]?.innerText.replace("Date Issued:", "")),
      submission_deadline: clean(metaEls[3]?.innerText.replace("Submission Deadline:", "")),
      category: clean(metaEls[4]?.innerText.replace("Category:", "")),
      sections
    };
  });

  await browser.close();
  return data;
}

File: scraper_py/scrape_tender.py
from playwright.sync_api import sync_playwright
import re


def scrape_tender(url: str) -> dict:
    with sync_playwright() as p:
        browser = p.chromium.launch(
            headless=True,
            args=["--no-sandbox", "--disable-setuid-sandbox"]
        )
        page = browser.new_page()
        page.goto(url, wait_until="networkidle")

        page.wait_for_selector(".doc-title")

        def clean(text):
            if not text:
                return None
            return re.sub(r"\s+", " ", text).strip()

        meta_elements = page.query_selector_all(".meta p")

        sections = {}
        for sec in page.query_selector_all("section"):
            h2 = sec.query_selector("h2")
            title = clean(h2.inner_text()) if h2 else None

            if title:
                content = clean(sec.inner_text().replace(title, ""))
                sections[title] = content

        data = {
            "project_name": clean(page.query_selector(".doc-title").inner_text()),
            "issued_by": clean(meta_elements[0].inner_text().replace("Issued By:", "")) if len(meta_elements) > 0 else None,
            "rfp_reference": clean(meta_elements[1].inner_text().replace("RFP Reference No.:", "")) if len(meta_elements) > 1 else None,
            "date_issued": clean(meta_elements[2].inner_text().replace("Date Issued:", "")) if len(meta_elements) > 2 else None,
            "submission_deadline": clean(meta_elements[3].inner_text().replace("Submission Deadline:", "")) if len(meta_elements) > 3 else None,
            "category": clean(meta_elements[4].inner_text().replace("Category:", "")) if len(meta_elements) > 4 else None,
            "sections": sections
        }

        browser.close()
        return data